RecoLocalMuon/GEMSegment/plugins/ME0SegAlgoMM.cc

```diff
@@ -0,0 +1,320 @@
+ME0SegAlgoMM::chainHits(const EnsembleHitContainer & rechits) {
+
+  ProtoSegments rechits_chains; 
+  EnsembleHitContainer temp;
+  ProtoSegments seeds;
+
+  std::vector <bool> usedCluster;
+
+  // split rechits into subvectors and return vector of vectors:
+  // Loop over rechits
+  // Create one seed per hit
+  for(unsigned int i = 0; i < rechits.size(); ++i) {
+    temp.clear();
+    temp.push_back(rechits[i]);
+    seeds.push_back(temp);
+    usedCluster.push_back(false);
+  }
+
+  // merge chains that are too close ("touch" each other)
+  for(size_t NNN = 0; NNN < seeds.size(); ++NNN) {
+    for(size_t MMM = NNN+1; MMM < seeds.size(); ++MMM) {
+      if(usedCluster[MMM] || usedCluster[NNN]){
+        continue;
+      }
+      // all is in the way we define "good";
+      // try not to "cluster" the hits but to "chain" them;
+      // it does the clustering but also does a better job
+      // for inclined tracks (not clustering them together;
+      // crossed tracks would be still clustered together) 
+      // 22.12.09: In fact it is not much more different 
+      // than the "clustering", we just introduce another
+      // variable in the game - Z. And it makes sense 
+      // to re-introduce Y (or actually wire group mumber)
+      // in a similar way as for the strip number - see
+      // the code below.
+      bool goodToMerge  = isGoodToMerge(seeds[NNN], seeds[MMM]);
+      if(goodToMerge){
+        // merge chains!
+        // merge by adding seed NNN to seed MMM and erasing seed NNN
+
+        // add seed NNN to MMM (lower to larger number)
+        seeds[MMM].insert(seeds[MMM].end(),seeds[NNN].begin(),seeds[NNN].end());
+
+        // mark seed NNN as used
+        usedCluster[NNN] = true;
+        // we have merged a seed (NNN) to the highter seed (MMM) - need to contimue to
+        // next seed (NNN+1)
+        break;
+      }
+
+    }
+  }
+
+  // hand over the final seeds to the output
+  // would be more elegant if we could do the above step with
+  // erasing the merged ones, rather than the
+
+  for(size_t NNN = 0; NNN < seeds.size(); ++NNN) {
+    if(usedCluster[NNN]) continue; //skip seeds that have been marked as used up in merging
+    rechits_chains.push_back(seeds[NNN]);
+  }
+
+  //***************************************************************
+
+  return rechits_chains;
+}
+
+bool ME0SegAlgoMM::isGoodToMerge(EnsembleHitContainer & newChain, EnsembleHitContainer & oldChain) {
+   for(size_t iRH_new = 0;iRH_new<newChain.size();++iRH_new){
+    int layer_new = newChain[iRH_new]->me0Id().layer();     
+    float phi_new = theEnsemble.first->toGlobal(newChain[iRH_new]->localPosition()).phi();
+    float eta_new = theEnsemble.first->toGlobal(newChain[iRH_new]->localPosition()).eta();
+    for(size_t iRH_old = 0;iRH_old<oldChain.size();++iRH_old){      
+      int layer_old = oldChain[iRH_old]->me0Id().layer();
+      float phi_old = theEnsemble.first->toGlobal(oldChain[iRH_old]->localPosition()).phi();
```

Comment:
depending on the size of newChain and oldChain and the time the "OK -> return true"  happens,
here it makes more sense to precompute in one for loop the layer, phi, eta in vectors and the have the nested loop over these simple int/floats.

Don't copy-paste ``` theEnsemble.first->toGlobal(oldChain[iRH_old]->localPosition()) ``` in multiple lines, use a GlobalPosition temp instead